mm2src/coins/solana/solana_token.rs

```diff
@@ -132,11 +182,167 @@ impl MmCoin for SolanaToken {
+
+        let fut = async move {
+            let rpc = coin
+                .rpc_client()
+                .await
+                .map_err(|e| WithdrawError::Transport(e.into_inner()))?;
+
+            // `to` can be either a Solana address, or a token address. We create
+            // `to_token_account` regardless to support the both cases.
+            let to = SolanaAddress::from_str(&req.to).map_err(|e| WithdrawError::InvalidAddress(e.to_string()))?;
+            let to_token_account = get_associated_token_address(&to, &token.protocol_info.mint_address);
+
+            let balance = token
+                .my_balance()
+                .compat()
+                .await
+                .map_err(|e| WithdrawError::Transport(e.to_string()))?;
+
+            let amount_u64 = if req.max {
+                balance.spendable.to_u64().ok_or_else(|| {
+                    MmError::new(WithdrawError::InternalError(format!(
+                        "Couldn't convert {} to u64.",
+                        balance.spendable
+                    )))
+                })?
+            } else {
+                let scale = BigDecimal::from(10u64.pow(token.protocol_info.decimals as u32));
+                let big_decimal = &req.amount * &scale;
+
+                big_decimal.to_u64().ok_or_else(|| {
+                    MmError::new(WithdrawError::InternalError(format!(
+                        "Couldn't convert {big_decimal} to u64."
+                    )))
+                })?
+            };
+
+            if amount_u64 == 0 {
+                return MmError::err(WithdrawError::AmountTooLow {
+                    amount: req.amount,
+                    threshold: token.min_tx_amount(),
+                });
+            }
+
+            let amount_decimal = lamports_to_big_decimal(amount_u64, token.protocol_info.decimals);
+            if balance.spendable < amount_decimal {
+                return MmError::err(WithdrawError::NotSufficientBalance {
+                    coin: token.ticker.to_owned(),
+                    available: balance.spendable,
+                    required: amount_decimal,
+                });
+            }
+
+            // Instructions:
+            //  - Create recipient address if missing.
+            //  - Transfer.
+            let mut instructions = Vec::new();
+
+            if let Err(e) = rpc.get_account(&to_token_account) {
+                if !e.kind.to_string().contains("AccountNotFound") {
+                    return MmError::err(WithdrawError::Transport(e.to_string()));
+                }
+
+                instructions.push(create_associated_token_account(
+                    &coin.address,
+                    &to,
+                    &token.protocol_info.mint_address,
+                    &spl_token_program::id(),
+                ));
+            };
+
+            let transfer_ix = spl_token_program::instruction::transfer_checked(
+                &spl_token_program::id(),
+                &coin.address,
```

Comment:
shouldn't this be `token.address`, will we use the associated token account (ATA) as the address of a token or the wallet address? I would say it makes sense to have the token address the wallet address so that APIs work the same and return right data, I don't think ATA is ever user facing. But something to think about as one choice between these will make more sense when adding solana token functionalities elsewhere in the code.